src/controllers/bulk/bulksupported.h

```diff
@@ -1,18 +1,20 @@
-// A list of supported USB bulk devices
-
 #pragma once
 
-typedef struct bulk_supported {
-    unsigned short vendor_id;
-    unsigned short product_id;
-    unsigned char in_epaddr;
-    unsigned char out_epaddr;
-    unsigned int interface_number;
-} bulk_supported_t;
+#include <cstdint>
+
+// A list of supported USB bulk devices
+
+struct bulk_supported_t {
+    std::uint16_t vendor_id;
+    std::uint16_t product_id;
+    std::uint8_t in_epaddr;
+    std::uint8_t out_epaddr;
+    std::uint8_t interface_number;
```

Comment:
where do you get that from? its uint8 in the relevant struct. https://github.com/libusb/libusb/blob/467b6a8896daea3d104958bf0887312c5d14d150/libusb/libusb.h#L749

Reply:
strange, everywhere else in libusb.h it is an int.

Reply:
You point to bInterfaceNumber which is a byte. Everywhere else it is used as int because of -1 for invalid.

Reply:
right, so `int16_t` or `std::optional<std::uint8_t>` or `int`? for the `int` you'll have to battle it out with @JoergAtGithub because [he requested it](https://github.com/mixxxdj/mixxx/pull/13622#discussion_r1747537727)

Reply:
I am against  `std::optional<std::uint8_t>` because we need to convert that to an `int` before calling libusb functions. 
int16_t suits here when it is not narrowed form int. 

Here it is stored as unsigned int which is even more confusing, and probably wrong. 
```
#if defined(__WINDOWS__) || defined(__APPLE__)
    m_interfaceNumber = pDevice->value.interface_number;
#endif
```
This may reveals some insides. 
https://github.com/libusb/hidapi/commit/bd6be4d83b36d0dcd5401fb335384586b066a2eb
Is the guard excluding Linux still correct?

Reply:
This seems to be a bug at least: 
https://github.com/mixxxdj/mixxx/blob/564a31845a8695cc7fcedbca0fe5c705a9c10bf5/src/controllers/bulk/bulkcontroller.cpp#L180C30-L180C57

Reply:
This is code for the USB Bulk devices, but you're referencing hidapi. This is an unrelated library.

Reply:
Here an unsigned int is narrowed to a signed int and the invalid -1 value is not considered: 
`if (m_interfaceNumber && libusb_kernel_driver_active(m_phandle, m_interfaceNumber) == 1) {`

Reply:
https://github.com/libusb/libusb/blob/467b6a8896daea3d104958bf0887312c5d14d150/libusb/core.c#L1823C50-L1823C67
The interface_numbers 0...31 are valid.

Reply:
I will not insist on uint8, but Invalid can not occur here. The sole source of information is our lookuptable. And you can't access an USB device without valid interface number. Adding -1 to an entry in our lookuptable would be a bug on it's own.

If we change it to int, libusb_kernel_driver_active will typecast it to uint8 internally anyway, so it shouldn't make a difference: https://github.com/libusb/libusb/blob/467b6a8896daea3d104958bf0887312c5d14d150/libusb/core.c#L2084-L2099

Reply:
My point in the original review was that the type for interface_number was changed from `unsigned int` to `std::uint32_t` and hardcoding the fixed size of 32 bit seemed misleading to me.

Reply:
I can confirm that. libusb is using uint8_t and int. I think we need to adjust our code to that. 
I have not yet understand Linux situation.

Reply:
We have the option to revert the change here and carry on, there was no user visible issue with the old types, was it? 

If we decide for a new type here we may adopt 
int, because the API will convert the given value to int anyway. What does the 0 below actually mean? Reading the code and the comment, it could be an unused value what we may want to indicate with -1 or is it actually interface 0? 

Where is the term "tweak" coming from and why is it referenced as value in the structure and not as tweak?

Reply:
The interface number along with 0 have been introduced in a bulk action here: 

https://github.com/mixxxdj/mixxx/commit/5e162f8c5ecc866c38385d82ee28ea89448c53c1

What does 0 mean? Why are all 0?

Reply:
> We have the option to revert the change here and carry on, there was no user visible issue with the old types, was it?

Not really. I can revert if that is seriously a condition to merge...

> What does the 0 below actually mean? Reading the code and the comment, it could be an unused value what we may want to indicate with -1 or is it actually interface 0?

Its the default. 0 on most devices, but not all (thus the table). 

> Where is the term "tweak" coming from and why is it referenced as value in the structure and not as tweak?

Because those are the nonstandard values we actually want to tweak. I can also call it "quirk", which is the terminology the linux kernel uses for this sort of thing.

> Why are all 0?

Because the only device that has a non-zero interface has not been added yet. 
https://github.com/acolombier/mixxx/blob/881a482f2000bf888fd9599922129981b00ac8aa/src/controllers/bulk/bulksupported.h#L13-L18

Reply:
It's just numbering of the interfaces. We currently just support bulk devices from the manufacturer Hercules. And Hercules always use the first interface for bulk and the second for audio. This order is completely up to the manufacturer.

Reply:
At least untill all numbers are known we need here int(-1) for invalid. see my other comment.

Reply:
> @acolombier may be able to share some more info on why detach kernel driver code is used on windows even though it doesn't work and why its not used on linux even though its at least possible.

This turned out to be the least minimal snippet of code to work on both Mac and Windows. 

> We have the option to revert the change here and carry on, there was no user visible issue with the old types, was it?

Not sure why we would revert a non-breaking issue here, but that would only impact Windows and Mac ability to use the S4 Mk3, on which effort of getting it officially supported by Mixxx has faded away, so I guess it doesn't really matter.
It is quite clear that this change was the happy path MVP to get Windows and Mac running, as some user was asking support for Mac on Zulip S4 thread. But I also don't own either and don't have the interest/money/time to do so and work on something comprehensive here.